src/main/com/yetanalytics/lrs/pedestal/interceptor/xapi/statements/attachment.cljc

```diff
@@ -133,145 +122,110 @@
+          #?(:clj (with-open [in input-stream]
+                    (slurp in :encoding "UTF-8"))
+             :cljs input-stream)
+          {:keys [headers payload]}
+          (decode-sig jws)
+          {:keys [alg _typ]}
+          headers]
+      (cond
+        (not (#{"RS256" "RS384" "RS512"}
+              alg))
+        (throw (ex-info "JWS Algorithm MUST be RS256, RS384, or RS512"
+                        {:type      ::invalid-signature-alg
+                         :jws       jws
+                         :statement statement}))
+        (not (ss/statements-immut-equal?
+              ;; TODO: make sure this doesn't trigger on substatement paths
+              (dissoc statement "attachments")
+              payload))
+        (throw (ex-info "Statement signature does not match statement"
+                        {:type      ::invalid-signature-mismatch
+                         :jws       jws
+                         :statement statement}))
+        :else ; If everything's good, return the multipart w/ a new input stream
+        (assoc multipart
+               :input-stream
+               #?(:clj (ByteArrayInputStream. (.getBytes jws "UTF-8"))
+                  :cljs jws))))))
 
 (defn sig?
   "Predicate, returns true if the given attachment object is a signature"
   [attachment-object]
   (= "http://adlnet.gov/expapi/attachments/signature"
      (get attachment-object "usageType")))
 
-(defn make-sha-multipart-pair
-  "Given the attachment object `att-obj`, return a pair of its SHA and the
-   multipart, or `nil` if `att-obj` is a fileURL."
-  [sig multi-parts att-obj]
-  ;; If we match...
-  (if-let [[sha mps] (find multi-parts (get att-obj "sha2"))]
-    (let [mp (first mps)]
-      ;; If it's a signature...
-      (if (sig? att-obj)
-        (if (= (:content-type mp)
-               (get att-obj "contentType")
-               "application/octet-stream")
-          ;; If the ctype is valid, validate the sig
-          [sha (validate-sig sig mp)]
-          ;; If that ctype was wrong, throw
-          (throw
-           (ex-info
-            "Statement signature attachment contentType must be application/octet-stream"
-            {:type                 ::invalid-signature-attachment-content-type
-             :attachment-object    att-obj
-             :attachment-multipart mp})))
-        ;; If not, return the sha and multipart
-        [sha mp]))
-    ;; If we don't, this better be a fileURL
-    (when-not (get att-obj "fileUrl")
-      (throw (ex-info "Invalid multipart format"
-                      {:type ::invalid-multipart-format})))))
+(defn multipart-map-dedupe
+  "Given a list of multiparts, make a map of vectors of them by xapi hash.
+  Removes duplicate multiparts."
+  [multiparts]
+  (reduce (fn [m {:keys [headers] :as multipart}]
+            (let [sha (get headers "X-Experience-API-Hash")]
+              (if-let [extant (get m sha)]
+                m
+                (assoc m sha multipart))))
```

Comment:
This assumes that attachments with different data _will_ have different SHAs, right?

Reply:
Yeah, even if it isn't mathematically true 🙈 . Unfortunately there isn't a stable way to tie an attachment to a specific referencing object in a statement batch really, so there isn't much to be done about it.

Reply:
In that case you may want to add that to the documentation.

Reply:
Oh so hash collisions aren't _that_ common with SHAs? From our past convos I got the impression that the SHA2 algorithm was busted and that collisions were the order of the day. 😆

Reply:
In general, yes we like to assume that impossible things will happen, it's just that in this instance the spec limits us.